src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,306 @@
+    }
+
+    void publish_internal(Aws::SNS::SNSClient *client, const std::string &resourceArn, const std::string &payload, reply_callback_t cb) {
+      Aws::SNS::Model::PublishRequest req;
+      req.SetMessage(Aws::String(payload));
+      req.SetTopicArn(Aws::String(resourceArn));
+
+      auto context = Aws::MakeShared<AsyncCallerContext>("PublishAllocationTag", cb, cct);
+      client->PublishAsync(req, publish_callback, context);
+    }
+
+    void publish_internal(Aws::Lambda::LambdaClient *client, const std::string &resourceArn, const std::string &payload, reply_callback_t cb) {
+      Aws::Lambda::Model::InvokeRequest invokeRequest;
+      Aws::Utils::ARN arn((Aws::String(resourceArn)));
+      invokeRequest.SetInvocationType(Aws::Lambda::Model::InvocationType::RequestResponse);
+      invokeRequest.SetLogType(Aws::Lambda::Model::LogType::Tail);
+      invokeRequest.SetFunctionName(arn.GetResource());
+      std::shared_ptr<Aws::IOStream> body = Aws::MakeShared<Aws::StringStream>("Simple Allocation Tag");
+      *body << payload;
+      invokeRequest.SetBody(body);
+      invokeRequest.SetContentType("application/json");
+      auto context = Aws::MakeShared<AsyncCallerContext>("InvokeAllocationTag", cb, cct);
+      client->InvokeAsync(invokeRequest, invoke_callback, context);
+    }
+
+//    Works on messages in the queque
+    void run() {
+      while (!stopped) {
+        const auto count = messageQueue.consume_all(
+                std::bind(&Manager::publish_internal_wrapper, this, std::placeholders::_1));
+        if (count == 0) {
+          std::this_thread::sleep_for(idle_time);
+        }
+      }
+    }
+
+    // used in the dtor for message cleanup
+    static void delete_message(const message_t *message) {
+      delete message;
+    }
+
+  public:
+    Manager(size_t _max_clients, size_t max_queue_size,unsigned idle_time_ms, CephContext *_cct) : messageQueue(max_queue_size),
+                                                                                                   clientList(_max_clients),
+                                                                                                   runner(&Manager::run, this),
+                                                                                                   stopped(false), cct(_cct),
+                                                                                                   idle_time(std::chrono::milliseconds(idle_time_ms)){
+          options = Aws::SDKOptions();
+          Aws::InitAPI(options);
+    }
+    // dtor wait for thread to stop
+    // then connection are cleaned-up
+    ~Manager() {
+      stopped = true;
+      runner.join();
+      messageQueue.consume_all(delete_message);
+      Aws::ShutdownAPI(options);
+    }
+
+
+    AwsClient connect(const std::string &accessKey,
+                      const std::string &accessSecret,
+                      const std::string &arn,
+                      const std::string &caPath,
```

Comment:
you can remove the `caPath` for now - please remove from all places (including doc)